src/session.rs

```diff
@@ -383,15 +404,22 @@ where S::Artifact: IntoInit<M::Init>
 
             if !act.is_empty() {
                 #[cfg(feature = "log")]
-                log::trace!(target: M::NAME, "Sending handshake act: {act:02x?}");
+                log::trace!(target: M::NAME, "Sending handshake act on write: {act:02x?}");
 
                 self.session.write_all(&act)?;
-            }
+            } else {
+                #[cfg(feature = "log")]
+                log::trace!(target: M::NAME, "Handshake complete, passing data to inner session");
 
-            return Err(io::ErrorKind::Interrupted.into());
+                return self.session.write(buf);
+            }
         }
 
-        self.session.write(buf)
+        if buf.is_empty() {
+            Ok(0)
+        } else {
+            Err(io::ErrorKind::Interrupted.into())
+        }
     }
```

Comment:
@cloudhead important!

Before, I was unconditionally passing to the inner session write operation. Now, I do that only if the handshake in the current session is complete

Reply:
Ah that means you can't write before the handshake anymore? That sounds good.

Reply:
Nope, you also was not being able to write before the handshake before...